StackOverflowError in FunctionDeclaration.equals

# Stack trace

```
18:03:57,633 INFO  MeasurementHolder VariableUsageResolver: Executing Pass done in 39 ms
18:03:57,634 INFO  MeasurementHolder CallResolver: Executing Pass
18:03:57,698 INFO  MeasurementHolder TranslationManager: Translation into full graph done in 921 ms
java.util.concurrent.ExecutionException: java.lang.StackOverflowError
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
	at de.fraunhofer.aisec.cpg_vis_neo4j.Application.call(Application.kt:559)
	at de.fraunhofer.aisec.cpg_vis_neo4j.Application.call(Application.kt:95)
	at picocli.CommandLine.executeUserObject(CommandLine.java:2041)
	at picocli.CommandLine.access$1500(CommandLine.java:148)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2461)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2453)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2415)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2273)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2417)
	at picocli.CommandLine.execute(CommandLine.java:2170)
	at de.fraunhofer.aisec.cpg_vis_neo4j.ApplicationKt.main(Application.kt:630)
Caused by: java.lang.StackOverflowError
	at java.base/java.lang.String.indexOf(String.java:2419)
	at java.base/java.lang.String.indexOf(String.java:2376)
	at java.base/java.net.URI.percentNormalizedComparison(URI.java:1926)
	at java.base/java.net.URI.equal(URI.java:1851)
	at java.base/java.net.URI.equals(URI.java:1515)
	at kotlin.jvm.internal.Intrinsics.areEqual(Intrinsics.java:169)
	at de.fraunhofer.aisec.cpg.sarif.PhysicalLocation$ArtifactLocation.equals(PhysicalLocation.kt:42)
	at kotlin.jvm.internal.Intrinsics.areEqual(Intrinsics.java:169)
	at de.fraunhofer.aisec.cpg.sarif.PhysicalLocation.equals(PhysicalLocation.kt:63)
	at kotlin.jvm.internal.Intrinsics.areEqual(Intrinsics.java:169)
	at de.fraunhofer.aisec.cpg.graph.Node.equals(Node.kt:333)
	at de.fraunhofer.aisec.cpg.graph.statements.Statement.equals(Statement.kt:60)
	at de.fraunhofer.aisec.cpg.graph.statements.expressions.Expression.equals(Expression.kt:256)
	at de.fraunhofer.aisec.cpg.graph.statements.expressions.BinaryOperator.equals(BinaryOperator.kt:200)
	at java.base/java.util.Objects.equals(Objects.java:64)
	at java.base/java.util.ArrayList.equalsArrayList(ArrayList.java:561)
```


# Unexpected behavior

When loading the two source file at once, the cpg will crash with a `StackOverflowError`.
The root cause seems to be the `FunctionDeclaration.equals`, which includes the function body in the comparison.
In this specific case, the function `foo` is recursive, so the body includes an `INVOCES` edge back to the `FunctionDeclaration`.
The cpg parses for every translation unit the header file again and creates in this case two `FunctionDeclaration` for the `foo` function. 
My assumption is, that because the two `FunctionDeclaration` are identical in all properties, but still two distinct objects and maybe connected in a circle, the `equals` will never end.
The error does not occur, when the class `A` is not a template class.
The error can also be circumvented, by using the unity-build mode, because in this case, then the header file is parsed only once and only one `FunctionDeclaration` for `foo` will be created.


# Code of interest

header.h
```c++
template<class K>
struct A {
    int foo(int i);
};

template<class K>
int A<K>::foo(int i) {
    return foo(i + 1);
}
```

main1.cpp
```c++
#include "header.h"

int main() {
    int i = 1;
    A<int> a;
    return a.foo(i);
}
```

main2.cpp
```c++
#include "header.h"

int main() {
    int i = 1;
    A<int> a;
    return a.foo(i);
}
```

# Extensions to the library, e.g. added or changed LanguageFrontends or Passes

None


# Changes to the configuration

None

--- a/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/declarations/FunctionDeclaration.kt
+++ b/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/declarations/FunctionDeclaration.kt
@@ -233,12 +233,10 @@ open class FunctionDeclaration : ValueDeclaration(), DeclarationHolder {
             body == other.body &&
             parameters == other.parameters &&
             propertyEqualsList(parameterEdges, other.parameterEdges) &&
-            throwsTypes == other.throwsTypes &&
-            overriddenBy == other.overriddenBy &&
-            overrides == other.overrides)
+            throwsTypes == other.throwsTypes)
     }
 
-    override fun hashCode() = Objects.hash(super.hashCode(), parameters, throwsTypes, overrides)
+    override fun hashCode() = Objects.hash(super.hashCode(), body, parameters, throwsTypes)
 
     override fun addDeclaration(declaration: Declaration) {
         if (declaration is ParamVariableDeclaration) {

--- a/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/statements/GotoStatement.kt
+++ b/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/statements/GotoStatement.kt
@@ -38,9 +38,7 @@ class GotoStatement : Statement() {
         if (other !is GotoStatement) {
             return false
         }
-        return super.equals(other) &&
-            labelName == other.labelName &&
-            targetLabel == other.targetLabel
+        return super.equals(other) && labelName == other.labelName
     }
 
     override fun hashCode() = Objects.hash(super.hashCode(), labelName, targetLabel)

--- a/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/statements/expressions/CallExpression.kt
+++ b/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/statements/expressions/CallExpression.kt
@@ -303,8 +303,6 @@ open class CallExpression : Expression(), HasType.TypeObserver, SecondaryTypeEdg
         return super.equals(other) &&
             arguments == other.arguments &&
             propertyEqualsList(argumentEdges, other.argumentEdges) &&
-            invokes == other.invokes &&
-            propertyEqualsList(invokeEdges, other.invokeEdges) &&
             templateParameters == other.templateParameters &&
             propertyEqualsList(templateParameterEdges, other.templateParameterEdges) &&
             templateInstantiation == other.templateInstantiation &&
@@ -312,8 +310,8 @@ open class CallExpression : Expression(), HasType.TypeObserver, SecondaryTypeEdg
     }
 
     // TODO: Not sure if we can add the template, templateParameters, templateInstantiation fields
-    // here
-    override fun hashCode() = Objects.hash(super.hashCode(), arguments, invokes)
+    //  here
+    override fun hashCode() = Objects.hash(super.hashCode(), arguments)
 
     override fun updateType(typeState: Collection<Type>) {
         for (t in typeTemplateParameters) {

--- a/cpg-language-cxx/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/cxx/CXXLanguageFrontendTest.kt
+++ b/cpg-language-cxx/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/cxx/CXXLanguageFrontendTest.kt
@@ -1592,4 +1592,22 @@ internal class CXXLanguageFrontendTest : BaseTest() {
         assertNotNull(ptr)
         assertLocalName("decltype(nullptr)", ptr.type)
     }
+
+    @Test
+    fun testRecursiveHeaderFunction() {
+        val file = File("src/test/resources/cxx/fix-1226")
+        val result =
+            analyze(
+                listOf(file.resolve("main1.cpp"), file.resolve("main2.cpp")),
+                file.toPath(),
+                true
+            )
+        assertNotNull(result)
+
+        // For now, we have duplicate functions because we include the header twice. This might
+        // change in the future. The important thing is that this gets parsed at all because we
+        // previously had a loop in our equals method
+        val functions = result.functions { it.name.localName == "foo" && it.isDefinition }
+        assertEquals(2, functions.size)
+    }
 }

--- a/cpg-language-cxx/src/test/resources/cxx/fix-1226/header.h
+++ b/cpg-language-cxx/src/test/resources/cxx/fix-1226/header.h
@@ -1,0 +1,9 @@
+template<class K>
+struct A {
+    int foo(int i);
+};
+
+template<class K>
+int A<K>::foo(int i) {
+    return foo(i + 1);
+}

--- a/cpg-language-cxx/src/test/resources/cxx/fix-1226/main1.cpp
+++ b/cpg-language-cxx/src/test/resources/cxx/fix-1226/main1.cpp
@@ -1,0 +1,7 @@
+#include "header.h"
+
+int main() {
+    int i = 1;
+    A<int> a;
+    return a.foo(i);
+}

--- a/cpg-language-cxx/src/test/resources/cxx/fix-1226/main2.cpp
+++ b/cpg-language-cxx/src/test/resources/cxx/fix-1226/main2.cpp
@@ -1,0 +1,7 @@
+#include "header.h"
+
+int main() {
+    int i = 1;
+    A<int> a;
+    return a.foo(i);
+}

